runtime/sqlite_record_logger.hpp

```diff
@@ -0,0 +1,205 @@
+
+/**
+ * There are many SQLite3 wrapper libraries.
+ * [List source](http://srombauts.github.io/SQLiteCpp/#see-also---some-other-simple-c-sqlite-wrappers)
+ * TODO: this
+ */
+
+namespace ILLIXR {
+
+class sqlite_thread {
+public:
+	sqlite3pp::database prep_db() {
+		if (!std::experimental::filesystem::exists(dir)) {
+			std::experimental::filesystem::create_directory(dir);
+		}
+
+		std::string path = dir / (table_name + std::string{".sqlite"});
+		return sqlite3pp::database{path.c_str()};
+	}
+
+	std::string prep_insert_str() {
+		std::string drop_table_string = std::string{"DROP TABLE IF EXISTS "} + table_name + std::string{";"};
+		db.execute(drop_table_string.c_str());
+
+		std::string create_table_string = std::string{"CREATE TABLE "} + table_name + std::string{"("};
+		for (unsigned i = 0; i < rh.get_columns(); ++i) {
+			create_table_string += rh.get_column_name(i) + std::string{" "};
+			if (false) {
+			} else if (rh.get_column_type(i) == typeid(std::size_t)) {
+				create_table_string += std::string{"INTEGER"};
+			} else if (rh.get_column_type(i) == typeid(bool)) {
+				create_table_string += std::string{"INTEGER"};
+			} else if (rh.get_column_type(i) == typeid(std::chrono::nanoseconds)) {
+				create_table_string += std::string{"INTEGER"};
+			} else if (rh.get_column_type(i) == typeid(std::chrono::high_resolution_clock::time_point)) {
+				create_table_string += std::string{"INTEGER"};
+			} else if (rh.get_column_type(i) == typeid(std::string)) {
+				create_table_string += std::string{"TEXT"};
+			} else if (rh.get_column_type(i) == typeid(double)) {
+				create_table_string += std::string{"REAL"};
+			} else {
+				throw std::runtime_error{std::string{"type "} + std::string{rh.get_column_type(i).name()} + std::string{" not found"}};
+			}
+			create_table_string += std::string{", "};
+		}
+		create_table_string.erase(create_table_string.size() - 2);
+		create_table_string += std::string{");"};
+		db.execute(create_table_string.c_str());
+
+		std::string insert_string = std::string{"INSERT INTO "} + table_name + std::string{" VALUES ("};
+		for (unsigned i = 0; i < rh.get_columns(); ++i) {
+			insert_string += std::string{"?"} + std::to_string(i+1) + std::string{", "};
+		}
+		insert_string.erase(insert_string.size() - 2);
+		insert_string += std::string{");"};
+		return insert_string;
+	}
+
+	sqlite_thread(const record_header& rh_)
+		: rh{rh_}
+		, table_name{rh.get_name()}
+		, db{prep_db()}
+		, insert_str{prep_insert_str()}
+		, insert_cmd{db, insert_str.c_str()}
+		, thread{std::bind(&sqlite_thread::pull_queue, this)}
+	{ }
+
+	void pull_queue() {
+		const std::size_t max_record_batch_size = 1024 * 256;
+		std::vector<record> record_batch {max_record_batch_size};
+		std::size_t actual_batch_size;
+
+		std::cout << "thread," << std::this_thread::get_id() << ",sqlite thread," << table_name << std::endl;
+
+		std::size_t processed = 0;
+		while (!terminate.load()) {
+			std::this_thread::sleep_for(std::chrono::seconds{1});
```

Comment:
Is it possible to do all of this _after_ terminate? That is, buffer everything until the very end. That will essentially ensure that there is no interference from the sqlite threads.

Reply:
It does delay everything until after termination, because the body of this while loop is commented out.

Reply:
I have no idea how I missed that.